// awilix-router-core/lib/controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var state_util_1 = require("./state-util");
var http_verbs_1 = require("./http-verbs");
var symbols_1 = require("./symbols");
/**
 * Configures routing config for a class or function to be invoked by a router.
 *
 * @example
 *   const api = ({ todoService }) => ({
 *     find: (ctx) => { ... }
 *   })
 *
 *   export default createController(api)
 *     .prefix('/todos')
 *     .before(bodyParser())
 *     .get('/', 'find')
 *     .get('/:id', 'get')
 *     .post('/:id', 'create', {
 *       before: [authenticate()],
 *     })
 *     .patch('/:id', 'update', {
 *       before: [authenticate()]
 *     })
 */
function createController(ClassOrFunction) {
    return createControllerFromState(ClassOrFunction, state_util_1.createState());
}
exports.createController = createController;
/**
 * Creates a builder from existing state.
 * This is used internally, but exported for convenience.
 *
 * @param ClassOrFunction The target to invoke.
 * @param state Existing state to continue building on.
 */
function createControllerFromState(ClassOrFunction, state) {
    var _a;
    var builder = (_a = {},
        _a[symbols_1.STATE] = state,
        _a[symbols_1.IS_CONTROLLER_BUILDER] = true,
        _a.target = ClassOrFunction,
        _a.get = createVerbFunction(http_verbs_1.HttpVerbs.GET),
        _a.post = createVerbFunction(http_verbs_1.HttpVerbs.POST),
        _a.put = createVerbFunction(http_verbs_1.HttpVerbs.PUT),
        _a.patch = createVerbFunction(http_verbs_1.HttpVerbs.PATCH),
        _a.delete = createVerbFunction(http_verbs_1.HttpVerbs.DELETE),
        _a.head = createVerbFunction(http_verbs_1.HttpVerbs.HEAD),
        _a.options = createVerbFunction(http_verbs_1.HttpVerbs.OPTIONS),
        _a.connect = createVerbFunction(http_verbs_1.HttpVerbs.CONNECT),
        _a.all = createVerbFunction(http_verbs_1.HttpVerbs.ALL),
        _a.prefix = function (path) {
            return createControllerFromState(ClassOrFunction, state_util_1.addRoute(state, null, path));
        },
        _a.before = function (middleware) {
            return createControllerFromState(ClassOrFunction, state_util_1.addBeforeMiddleware(state, null, middleware));
        },
        _a.after = function (middleware) {
            return createControllerFromState(ClassOrFunction, state_util_1.addAfterMiddleware(state, null, middleware));
        },
        _a.verbs = function (verbs, path, method, opts) {
            state = state_util_1.addRoute(state, method, path);
            state = state_util_1.addHttpVerbs(state, method, verbs);
            if (opts) {
                if (opts.before) {
                    state = state_util_1.addBeforeMiddleware(state, method, opts.before);
                }
                if (opts.after) {
                    state = state_util_1.addAfterMiddleware(state, method, opts.after);
                }
            }
            return createControllerFromState(ClassOrFunction, state);
        },
        _a);
    return builder;
    /**
     * Creates a preconfigured verb function.
     *
     * @param verb
     */
    function createVerbFunction(verb) {
        return function configureRoute(path, method, opts) {
            return builder.verbs([verb], path, method, opts);
        };
    }
}
exports.createControllerFromState = createControllerFromState;
